source/visualisation/solution_monitoring_inverseproblem.py

```diff
@@ -0,0 +1,359 @@
+import sys
+
+import numpy as np
+import pandas as pd
+import igraph
+import matplotlib.pyplot as plt
+from matplotlib.ticker import MaxNLocator
+
+import source.flow_network as flow_network
+import source.inverse_model as inverse_model
+from types import MappingProxyType
+
+
+class SolutionMonitoring(object):
+
+    def __init__(self, flownetwork: flow_network.FlowNetwork, inversemodel: inverse_model.InverseModel,
+                 PARAMETERS: MappingProxyType):
+
+        self.flownetwork = flownetwork
+        self.inversemodel = inversemodel
+        self._PARAMETERS = PARAMETERS
+
+    def get_arrays_for_plots(self):
+
+        current_iteration = self.inversemodel._current_iteration
+
+        # Create arrays for plotting cost_function vs iterations
+        f_h = self.inversemodel.f_h
+        self.inversemodel._iteration_array = np.append(self.inversemodel._iteration_array, current_iteration)
+        self.inversemodel._f_h_array = np.append(self.inversemodel._f_h_array, f_h)
+
+        # Create arrays for plotting tuning value - Plot: tuning_value vs iterations
+        edge_id_target = self.inversemodel.edge_constraint_eid
+        edge_constraint_range = self.inversemodel.edge_constraint_range_pm
+        edge_constraint_type = self.inversemodel.edge_constraint_type
+
+        flow_rate = self.flownetwork.flow_rate
+        rbc_velocity = self.flownetwork.rbc_velocity
+
+        # Only for precision measurements, so that the constraint range (edge_tar_range_pm) is zero  - target value
+        # constraint type: 1: Flow rate constraint & 2: Velocity constraint
+        is_target_type_1 = np.logical_and(edge_constraint_type == 1, edge_constraint_range == 0.)
+        is_target_type_2 = np.logical_and(edge_constraint_type == 2, edge_constraint_range == 0.)
+        current_flow_rate = flow_rate[edge_id_target[is_target_type_1]]
+        current_rbc_velocity = rbc_velocity[edge_id_target[is_target_type_2]]
+
+        if current_iteration == 0:
+            self.inversemodel._flow_rate_sim_array = np.array(current_flow_rate)
+            self.inversemodel._rbc_velocity_sim_array = np.array(current_rbc_velocity)
+        else:
+            self.inversemodel._flow_rate_sim_array = np.vstack((self.inversemodel._flow_rate_sim_array,
```

Comment:
Question: Is it a problem that these "dense" arrays can become very large? For example, if you iterate for 100'000 iteration steps and with many constraints, it might increase the computation time significantly, since in every iteration the array has to be modified with vstack. This always requires to copy the old array and adding a new one..

Maybe, an other option would be just to save csv files and then use a separate script to plot the convergence.

Just keep this in mind, but I'm a bit worried that for the large realistic networks you would loose a lot of time because of using the vstack function at late iteration steps. In that case, all our efforts to speed up the code would not be relevant any more..

--> you can also try this in a simple python script with a for-loop and for example check how long it takes to append new values to an existing array with vstack, if you use up to 200'000 iterations.. maybe, it's not a problem at all..

Reply:
Yes, this is a very good point! I will definitely check it by using a simple example. 

So do you think it would be better to use this 'SolutionMonitoring' class only to store the data for each iteration and then plot the corresponding graphs in a different script independently?

By doing that, the SolutionMonitoring class will not affect the speed of the inverse_model python script. (Right?)
This makes more sense.

Reply:
"So do you think it would be better to use this 'SolutionMonitoring' class only to store the data for each iteration and then plot the corresponding graphs in a different script independently?

By doing that, the SolutionMonitoring class will not affect the speed of the inverse_model python script. (Right?)
This makes more sense."
--> yes, I believe this would be better